app/models/pipeline_run.rb

```diff
@@ -153,38 +243,311 @@ def retry
+    # those outputs should be marked as failed. Otherwise we will never
+    # stop checking for them.
+    # [ TODO: move the check on "finalized" (column managed by pipeline_monitor)
+    #   to an S3 interface in order to give us the option of running pipeline_monitor
+    #   in a new environment that result_monitor does not have access to.
+    # ]
+    output = output_state.output
+    state = output_state.state
+    if finalized? && !output_ready?(output)
+      output_state.update(state: STATUS_FAILED)
+      return
+    end
+    # Otherwise, if the output is ready and has not yet been loaded or enqueued
+    # for loading, enqueue a loader job.
+    if output_ready?(output) && state == STATUS_UNKNOWN
+      output_state.update(state: STATUS_LOADING_QUEUED)
+      Resque.enqueue(ResultMonitorLoader, id, output)
+    end
+  end
+
+  def handle_success
+    # Check if this was the last run in a project and act accordingly:
+    if sample.project.results_complete?
+      notify_users
+      sample.project.create_or_update_project_background if sample.project.background_flag == 1
+    end
+  end
+
+  def load_stats_file
+    # TODO: make it a single file rather than using stage-specific.
+    host_filtering_job_stats_s3 = "#{host_filter_output_s3_path}/#{STATS_JSON_NAME}"
+    alignment_job_stats_s3 = "#{alignment_output_s3_path}/#{STATS_JSON_NAME}"
+    if file_generated_since_jobstats?(alignment_job_stats_s3)
+      load_job_stats(alignment_job_stats_s3)
+    elsif file_generated_since_jobstats?(host_filtering_job_stats_s3)
+      load_job_stats(host_filtering_job_stats_s3)
+    end
+  end
+
+  def skip_host_filtering?
+    sample.host_genome && sample.host_genome.name == HostGenome::NO_HOST_NAME
+  end
+
+  def all_output_states_terminal?
+    output_states.pluck(:state).all? { |s| [STATUS_LOADED, STATUS_FAILED].include?(s) }
+  end
+
+  def monitor_results
+    return if results_finalized?
+
+    # Get pipeline_version, which determines S3 locations of output files.
+    # If pipeline version is not present, we cannot load results yet.
+    # [ We use "file_generated_since_run(pipeline_version_file)" because "pipeline_version_file"
+    #   is not in the versioned result folder, so it gets overwritten with each new run.
+    #   TODO: change that, so that we can get rid of "file_generated_since_run".
+    # ]
+    if pipeline_version.blank? && file_generated_since_run(pipeline_version_file)
+      update(pipeline_version: fetch_pipeline_version)
+    end
+    return if pipeline_version.blank?
+
+    # Load any new outputs that have become available:
+    output_states.each do |o|
+      check_and_enqueue(o)
+    end
+
+    # Update job stats:
+    load_stats_file
+
+    # Check if run is complete:
+    if output_states.pluck(:state).all? { |s| s == STATUS_LOADED }
+      update(results_finalized: FINALIZED_SUCCESS)
+      handle_success
+    elsif all_output_states_terminal?
+      update(results_finalized: FINALIZED_FAIL)
+    end
```

Comment:
This is fine, but would probably be more efficient to not repeat the query in the common case (polling).
```
if all_output_states_terminal?
    if all_output_states_loaded?
        update(results_finalized: FINALIZED_SUCCESS)
        handle_success
    else
        update(results_finalized: FINALIZED_FAIL)
    end
end
```

Reply:
ok changed